restapi/src/test/java/io/stargate/web/docsapi/service/QueryExecutorTest.java

```diff
@@ -0,0 +1,426 @@
+
+    List<RawDocument> r1 = get(executor.queryDocs(allDocsQuery, pageSize, null));
+    assertThat(r1).extracting(RawDocument::id).containsExactly("1", "2", "3", "4", "5");
+
+    assertThat(r1.get(0).hasPagingState()).isTrue();
+    ByteBuffer ps1 = r1.get(0).makePagingState();
+    List<RawDocument> r2 = get(executor.queryDocs(allDocsQuery, pageSize, ps1));
+    assertThat(r2).extracting(RawDocument::id).containsExactly("2", "3", "4", "5");
+
+    assertThat(r1.get(1).hasPagingState()).isTrue();
+    ByteBuffer ps2 = r1.get(1).makePagingState();
+    List<RawDocument> r3 = get(executor.queryDocs(allDocsQuery, pageSize, ps2));
+    assertThat(r3).extracting(RawDocument::id).containsExactly("3", "4", "5");
+
+    assertThat(r1.get(3).hasPagingState()).isTrue();
+    ByteBuffer ps4 = r1.get(3).makePagingState();
+    List<RawDocument> r4 = get(executor.queryDocs(allDocsQuery, pageSize, ps4));
+    assertThat(r4).extracting(RawDocument::id).containsExactly("5");
+  }
+
+  @ParameterizedTest
+  @CsvSource({"1", "100", "5000"})
+  void testFullScanDeep(int pageSize) {
+    int N = 100_000;
+    Builder<Map<String, Object>> rows = ImmutableList.builder();
+    for (int i = 0; i < N; i++) { // generate 10 pages of data
+      rows.add(row("" + i, "a", 11.0d)); // one row per document
+    }
+
+    withQuery(table, "SELECT * FROM %s").withPageSize(pageSize).returning(rows.build());
+
+    // Testing potential stack overflow in Rx pipelines
+    assertThat(get(executor.queryDocs(allDocsQuery, pageSize, null))).hasSize(N);
+  }
+
+  @ParameterizedTest
+  @CsvSource({"4", "10", "20", "50", "100", "500", "1000", "5000"})
+  void testPartialFullScan(int pageSize) throws InterruptedException {
+    Builder<Map<String, Object>> rows = ImmutableList.builder();
+    for (int i = 0; i <= 10 * pageSize; i++) { // generate 10 pages of data
+      rows.add(row("" + i, "a", 11.0d)); // one row per document
+    }
+
+    QueryAssert queryAssert =
+        withQuery(table, "SELECT * FROM %s").withPageSize(pageSize).returning(rows.build());
+
+    Flowable<RawDocument> flowable = executor.queryDocs(allDocsQuery, pageSize, null);
+    AtomicReference<Subscription> subscription = new AtomicReference<>();
+    Semaphore sem = new Semaphore(0);
+    flowable.subscribe(
+        new Subscriber<RawDocument>() {
+          @Override
+          public void onSubscribe(Subscription s) {
+            subscription.set(s);
+          }
+
+          @Override
+          public void onNext(RawDocument rawDocument) {
+            sem.release();
+          }
+
+          @Override
+          public void onError(Throwable t) {
+            fail("Rx error: " + t, t);
+          }
+
+          @Override
+          public void onComplete() {
+            // nop
+          }
+        });
+
+    assertThat(subscription.get()).isNotNull();
+
+    subscription.get().request(1);
+    assertThat(sem.tryAcquire(30, TimeUnit.SECONDS)).isTrue();
+    queryAssert.assertExecuteCount().isEqualTo(2);
```

Comment:
so this is very strange for me, we request 1 element, but are already on the page 2, although we should actually be fine with loading only one page.. I observed same behavior with the updates I proposed, I think test is wrong here? Wdyt?

Reply:
Rx code subscribes to the "next" element on emitting "current". If our execution is driven from subscription callbacks we end up "proactively" executing page requests.

Reply:
I was able to achieve the only one hit with:
* `limit(1)` to the documents flowable
* using `TestSubscriber.create(1)` which means take only one 

So at least I am happy now, having a second RS initialized on first emitting is also not bad for the performance imo and should stay like that in fact, even if this sends some extra queries to the DB..

Reply:
Fixed / reworked in the latest push. Please review again.